stream-parent/stream-client/src/main/java/org/interledger/stream/sender/SimpleStreamSender.java

```diff
@@ -804,28 +835,90 @@ void handleReject(
+            originalPreparePacket, originalStreamPacket, rejectPacket
+        );
+      } else { // if (ErrorFamily.TEMPORARY.equals(rejectPacket.getCode().getErrorFamily())) {
+        logger.warn(
+            "Temporary ILPv4 transport outage. originalPreparePacket={} originalStreamPacket={} rejectPacket={}",
+            originalPreparePacket, originalStreamPacket, rejectPacket
+        );
+      }
+    }
 
-          } else {
-            logger.error(
-                "Encountered Final ILPv4 error. Retrying, but this sendMoney will likely hang until timeout."
-                    + " originalPreparePacket={} originalStreamPacket={} rejectPacket={}",
-                originalPreparePacket, originalStreamPacket, rejectPacket);
-          }
-          break;
+    /**
+     * <p>A helper method to centralize all logic around when to trigger an "unrecoverable error" that will stop this
+     * sender from schedule further packets.</p>
+     *
+     * <p>This method functions according to the following rules:</p>
+     *
+     * <p>Interledger Reject packets with the following attributes will immediately trigger an UNRECOVERABLE_ERROR
+     * condition, stopping the sendMoney operation:</p>
+     * <ol>
+     *   <li>Any F-family rejections (except F08 and F99, which are used for congestion control and rate-probing).</li>
+     *   <li>Any R-family rejections.</li>
+     * </ol>
+     *
+     * <p>Interledger Reject packets with the following attributes MAY trigger an UNRECOVERABLE_ERROR condition after
+     * some threshold (this logic is not yet implemented, and may never be):</p>
+     * <ol>
+     *  <li>T00 Errors (generally should be accepted up to some threshold)</li>
+     *  <li>T01 Errors (generally should be accepted up to some threshold)</li>
+     *  <li>F99 Errors (generally should be accepted up to some threshold; used for FX problems where prepare amount is
+     *  less than min amount the receiver should receive).</li>
+     * </ol>
+     *
+     * <p>Interledger Reject packets with the following attributes will NEVER trigger an UNRECOVERABLE_ERROR condition:</p>
+     * <ol>
+     *  <li>T02-T99 Errors (these may resolve with time)</li>
+     *  <li>F08 Errors (used for congestion control)</li>
+     * </ol>
+     *
+     * @param rejectPacket An {@link InterledgerRejectPacket} that was received from an outgoing {@link Link} in
+     *                     response to a prepare packet.
+     */
+    @VisibleForTesting
+    void checkForAndTriggerUnrecoverableError(final InterledgerRejectPacket rejectPacket) {
+      Objects.requireNonNull(rejectPacket);
+
+      if (ErrorFamily.FINAL.equals(rejectPacket.getCode().getErrorFamily())) {
+        // Most Final errors trigger immediate stoppage of send operations.
+        if (!rejectPacket.getCode().equals(F08_AMOUNT_TOO_LARGE) &&
+            !rejectPacket.getCode().equals(F99_APPLICATION_ERROR)) {
```

Comment:
Seeing these 2 error codes repeated. Might make sense to create `static final List<InterledgerErrorCode> retryableFinalErrors = Lists.newArrayList(F08, F99)` and then check if `retryableFinalErrors.contains(rejectPacket.getCode())`

Reply:
This is fixed.